### ISSUE TITLE
[Feature Request] Support for Experimental Codecs

### ISSUE THREAD

**Issue by sarthakaggarwal97** [opensearch-project/OpenSearch#13723]
### Is your feature request related to a problem? Please describe

With the introduction of new custom codecs, it gets essential to bake the new codecs before marking them ready for production. This helps us to avoid potential index corruption risks, and allows for the community feedback to flow in before we make the codecs as generally available.

Earlier, we used to have sandbox for codecs OpenSearch, with the introduction of the separate repository for custom codecs, it gets important to have a similar functionality available in the codecs plugin as well.

### Describe the solution you'd like

We can extend the existing CodecSettings interface to know whether the codec is experimental upon index creation.

### Related component

Indexing

### Additional context

Coming from https://github.com/opensearch-project/custom-codecs/issues/148

**Comment by andrross** [comment]
@sarthakaggarwal97 Could the custom-codecs plugin itself create an register a static setting like `codecs.experimental.enabled` and only register the new codec if that setting is explicitly set to true?

**Comment by sarthakaggarwal97** [comment]
@andrross Do you mean a static cluster setting? It would be like a feature flag right. Please correct me if I am wrong.

We might have to utilize this interface so that we can mark codecs as experimental. Another benefit of an interface change is that it gives a way for other plugins as well to use it. We have plugins like Security Analytics, K-nn who also currently maintain their own codecs, and might have similar use cases in the future. 

I feel it will be a combination of both the setting and a clean way to mark the new codecs as experimental. We can then have assertions in place to block the create index call.

**Comment by andrross** [comment]
@sarthakaggarwal97 Yeah it would functionally work the same way as a feature flag, but just be scoped to the custom-codecs plugin in this case. I think in order to determine exactly _how_ we implement this feature we'll want to clearly define the user experience. If we extend the CodecSettings interface, how do we implement the behavior? What's a system admin's experience for enabling and disabling experimental codecs?

**Comment by sarthakaggarwal97** [comment]
@andrross since the validation to select the codec happens in [EngineConfig](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/EngineConfig.java#L133), we might need to implement the feature flag within OpenSearch itself. CodecSettings interface would help us know whether the particular codec is experimental or not, while the feature flag would make experimental codecs available via CodecService.

Tagging @reta @mgodwan @backslasht for inputs.

**Comment by reta** [comment]
> @andrross since the validation to select the codec happens in [EngineConfig](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/EngineConfig.java#L133), we might need to implement the feature flag within OpenSearch itself. CodecSettings interface would help us know whether the particular codec is experimental or not, while the feature flag would make experimental codecs available via CodecService.

I would :100: agree with @andrross that experimental codecs should be contained in `custom-codecs` plugin, including the experimental support. The `CodecSettings` is optional and we should not expect any codec to implement it (vs Apache Lucene's `Codec`).

**Comment by sarthakaggarwal97** [comment]
I guess we should be able to have the setting in custom-codecs. We might have to still expose the ```experimental``` via the CodecSettings interface. If would not validate the experimental nature of codecs in the EngineConfig, and the codecs are not available in CodecService (because they are experimental), the shards would fail.

Apologies if I am missing something here.

**Comment by reta** [comment]
> Apologies if I am missing something here, please correct me if I am wrong.

I would encourage to not rely on CodecSettings anyhow

**Comment by andrross** [comment]
> If we would not validate the experimental nature of codecs in the EngineConfig, and the codecs are not available in CodecService (because they are experimental), the shards would fail.

@sarthakaggarwal97 Can you elaborate a bit on what you mean here? I get that there are sharp edges on the user experience (i.e. if you disable an experimental codec then any existing indexes using that codec will fail), but that is sort of expected for these opt-in experimental features until they are stabilized and fully supported. To me, the simplest option here is to just let custom-codecs choose whether to register experimental codecs at startup based on the existence of a static setting. What about that implementation doesn't work with the experience you have in mind?

**Comment by sarthakaggarwal97** [comment]
@andrross in EngineConfig when the validation happens upon index creation, it loads the codec from NamedSPI itself. The codecs are registered with NamedSPI through [resources](https://github.com/opensearch-project/custom-codecs/blob/main/src/main/resources/META-INF/services/org.apache.lucene.codecs.Codec). 

Is it possible to load a different resource file based on an experimental settings?

**Comment by reta** [comment]
> Is it possible to load a different resource file based on an experimental settings?

It is not possible, the service loader mechanism does not support that (afaik)

**Comment by sarthakaggarwal97** [comment]
yeah, even I am not aware.
@andrross @reta what do you guys suggest in this case?

**Comment by sarthakaggarwal97** [comment]
The challenge would be make the experimental codecs _not available_ in the index.codec settings validations.

**Comment by reta** [comment]
> If the codecs are registered by NamedSPI, they would be available over [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/EngineConfig.java#L146)

Same answer as for `CodecSettings`, the `CodecAliases` is our and any attempts to use it (including validating settings) would fail

**Comment by peternied** [comment]
[Triage - attendees [1](https://github.com/srikanthpadakanti) [2](https://github.com/peternied) [3](https://github.com/lukas-vlcek) [4](https://github.com/jed326) [5](https://github.com/reta) [6](https://github.com/andrross) [7](https://github.com/kotwanikunal)]
@sarthakaggarwal97 Thanks for creating this issue
